Implementation of PushAVTransport and PushAVStreamTransportDelegate

raveendra-karu opened the pull request:
####  Changes
This PR implements the following :
- PushAV Stream Transport Delegate Implementation
- PushAV Transport creation and management
- PushAV commands handling

This PR depends on PushAV server cluster: https://github.com/project-chip/connectedhomeip/pull/37787
 
#### Testing
Push AV server cluster  test scripts are added: https://github.com/project-chip/connectedhomeip/pull/39727

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=38331) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=38331) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: raveendra-karu<br/>:x: chayan-a<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/project-chip/connectedhomeip?pullRequest=38331) it.</sub>

yufengwangca commented on examples/camera-app/linux/BUILD.gn:
```diff
@@ -56,10 +58,12 @@ executable("chip-camera-app") {
     "${chip_root}/examples/camera-app/linux/src/clusters/webrtc_provider/webrtc-provider-manager.cpp",
     "${chip_root}/examples/camera-app/linux/src/media-controller/media-controller.cpp",
     "${chip_root}/examples/camera-app/linux/src/network-stream-source/network-stream-source.cpp",
+    "${chip_root}/examples/camera-app/linux/src/pushav-transport.cpp",
```
```suggestion
    "${chip_root}/examples/camera-app/linux/src/pushav-transport/pushav-transport.cpp",
```

yufengwangca commented on examples/camera-app/linux/include/clusters/push_av_transport/push-av-transport-manager.h:
```diff
@@ -0,0 +1,88 @@
+
+private:
+    PushAVTransport * Transports[MAX_PUSH_TRANSPORT_CONNECTION_ID];                         // map for the transport objects
+    std::optional<TransportOptionsStruct> ConnectionsMap[MAX_PUSH_TRANSPORT_CONNECTION_ID]; // map for the transport options
```
Using fixed-size arrays (Transports and ConnectionsMap) for 65k entries is wasteful. Most entries will be unused, leading to excessive memory consumption.

raveendra-karu replied:
Refactored the code

yufengwangca commented on examples/camera-app/linux/include/clusters/push_av_transport/push-av-transport-manager.h:
```diff
@@ -0,0 +1,88 @@
+public:
+    PushAVTransportManager() { std::fill_n(Transports, MAX_PUSH_TRANSPORT_CONNECTION_ID, nullptr); }
+
+    ~PushAVTransportManager() {}
```
The destructor ~PushAVTransportManager() does not free allocated PushAVTransport objects.

raveendra-karu replied:
Done

yufengwangca commented on examples/camera-app/linux/src/clusters/push_av_transport/push-av-transport-manager.cpp:
```diff
@@ -0,0 +1,168 @@
+    }
+    else
+    {
+        Transports[connectionID]->~PushAVTransport();
```
DeallocatePushTransport incorrectly calls the destructor directly. Use delete instead:

delete Transports[connectionID];
Transports[connectionID] = nullptr;

raveendra-karu replied:
done

yufengwangca commented on examples/camera-app/linux/src/clusters/push_av_transport/push-av-transport-manager.cpp:
```diff
@@ -0,0 +1,168 @@
+Protocols::InteractionModel::Status PushAVTransportManager::ModifyPushTransport(const uint16_t connectionID,
+                                                                                const TransportOptionsStruct & outTransportOptions)
+{
+    ChipLogProgress(NotSpecified, "ModifyPushTransport manager called");
```
ModifyPushTransport updates the transport object but not ConnectionsMap, causing inconsistency.

ConnectionsMap[connectionID] = outTransportOptions; // Update stored options

yufengwangca commented on examples/camera-app/linux/src/clusters/push_av_transport/push-av-transport-manager.cpp:
```diff
@@ -0,0 +1,168 @@
+    }
+    else
+    {
+        Transports[connectionID]->mTransportTriggerType = outTransportOptions.triggerOptions.triggerType;
```
outTransportOptions is  out param, what has been modified within this function?

raveendra-karu replied:
Configuration need to be maintained for later use

yufengwangca commented on examples/camera-app/linux/include/clusters/push_av_transport/push-av-transport-manager.h:
```diff
@@ -0,0 +1,88 @@
+    CHIP_ERROR PersistentAttributesLoadedCallback() override;
+
+private:
+    PushAVTransport * Transports[MAX_PUSH_TRANSPORT_CONNECTION_ID];                         // map for the transport objects
```
The PushAVTransport (as a std::unique_ptr) and its options needed to be bundled, such as:

struct TransportEntry
{
    std::unique_ptr<PushAVTransport> transport;
    TransportOptionsStruct           options;
};

raveendra-karu replied:
handle this one with map

yufengwangca commented on src/app/clusters/push-av-stream-transport-server/push-av-stream-transport-server.h:
```diff
@@ -0,0 +1,266 @@
+     *
+     *   @param connectionID  [in]     Indicates the allocated connectionID to get the Stream Options Configuration of.
+     *
+     *   @param outtransportConfigurations  [out]     Single item list of mapped transport configuration or list if connectionID is
```
```suggestion
     *   @param outTransportConfigurations  [out]     Single item list of mapped transport configuration or list if connectionID is
```

yufengwangca replied:
The description states that passing a NULL connectionID should return a list of configurations. However, the code returns Status::Failure in this case. This is a mismatch.

yufengwangca commented on src/app/clusters/push-av-stream-transport-server/push-av-stream-transport-server.h:
```diff
@@ -0,0 +1,266 @@
+     */
+    virtual Protocols::InteractionModel::Status
+    FindTransport(const Optional<DataModel::Nullable<uint16_t>> & connectionID,
+                  DataModel::List<const TransportConfigurationStruct> & outtransportConfigurations) = 0;
```
```suggestion
                  DataModel::List<const TransportConfigurationStruct> & outTransportConfigurations) = 0;
```

raveendra-karu replied:
Done

yufengwangca commented on examples/camera-app/linux/include/clusters/push_av_transport/push-av-transport-manager.h:
```diff
@@ -0,0 +1,93 @@
+
+    Protocols::InteractionModel::Status
+    FindTransport(const Optional<DataModel::Nullable<uint16_t>> & connectionID,
+                  DataModel::List<const TransportConfigurationStruct> & outtransportConfigurations) override;
```
```suggestion
                  DataModel::List<const TransportConfigurationStruct> & outTransportConfigurations) override;
```

yufengwangca commented on examples/camera-app/linux/src/clusters/push_av_transport/push-av-transport-manager.cpp:
```diff
@@ -0,0 +1,209 @@
+        return Status::Failure;
+    }
+
+    std::vector<TransportConfigurationStruct> configList;
```
You build a local std::vector<…> configList, then wrap it in a DataModel::List that stores raw pointers into that vector. Once the function returns the vector is destroyed ➜ dangling pointers.

yufengwangca commented on examples/camera-app/linux/src/clusters/push_av_transport/push-av-transport-manager.cpp:
```diff
@@ -0,0 +1,209 @@
+PushAvStreamTransportManager::FindTransport(const Optional<DataModel::Nullable<uint16_t>> & connectionID,
+                                            DataModel::List<const TransportConfigurationStruct> & outtransportConfigurations)
+{
+    if (connectionID.Value().IsNull())
```
The code does not check if the Optional connectionID has a value before calling Value(). If the Optional is empty (no value), accessing Value() leads to undefined behavior.

raveendra-karu replied:
Done

yufengwangca commented on examples/camera-app/linux/src/clusters/push_av_transport/push-av-transport-manager.cpp:
```diff
@@ -0,0 +1,209 @@
+    return Status::Success;
+}
+
+Protocols::InteractionModel::Status PushAvStreamTransportManager::SetTransportStatus(const std::vector<uint16_t> connectionIDList,
```
```suggestion
Protocols::InteractionModel::Status PushAvStreamTransportManager::SetTransportStatus(const std::vector<uint16_t> & connectionIDList,
```

raveendra-karu replied:
Done

yufengwangca commented on src/app/clusters/push-av-stream-transport-server/push-av-stream-transport-server.h:
```diff
@@ -0,0 +1,266 @@
+     *   @return Success if the stream transport status is successfully set; otherwise, the command SHALL be rejected with an
+     * appropriate error.
+     */
+    virtual Protocols::InteractionModel::Status SetTransportStatus(const std::vector<uint16_t> connectionIDList, TransportStatusEnum transportStatus) = 0;
```
```suggestion
    virtual Protocols::InteractionModel::Status SetTransportStatus(const std::vector<uint16_t> & connectionIDList, TransportStatusEnum transportStatus) = 0;
```

raveendra-karu replied:
Done

yufengwangca commented:
This PR adds the server-side implementation of push-av-transport-manager. This part already pending under Boris’s review (https://github.com/project-chip/connectedhomeip/pull/37787), and he has put a group reject.

So we should either:

1. Wait for #37787 to be merged, or
2. Split out just the server implementation here and ask Boris to review it separately.

raveendra-karu commented on examples/camera-app/linux/src/clusters/push_av_transport/push-av-transport-manager.cpp:
```diff
@@ -0,0 +1,209 @@
+{
+    // Unregister all transports from Media Controller before deleting them. This will ensure that any ongoing streams are stopped.
+    for (auto & transport : mTransportMap)
+    {
```
Done

raveendra-karu committed:
Implement PushAVTransport and PushAVStreamTransportDelegate.

Co-Author: Tushant Yadav <t.yadav@samsung.com>
Co-Author: Chayan Ailawalia <c.ailawalia@samsung.com>

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

raveendra-karu committed:
addressed build issue

Co-Author: Tushant Yadav <t.yadav@samsung.com>

raveendra-karu committed:
Implemented delegate API

Co-Author: Tushant Yadav <t.yadav@samsung.com>

raveendra-karu committed:
restyled and commented cluster dependent code

Co-Author: Tushant Yadav <t.yadav@samsung.com>

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

raveendra-karu committed:
code cleanup

Co-Author: Tushant Yadav <t.yadav@samsung.com>

raveendra-karu committed:
Implement PushAVTransport integration and code cleanup

- Add PushAVTransport cluster support to camera-app
- Remove TODO comments as dependencies are now resolved
- Clean up BUILD.gn files and remove outdated comments

Author: Tushant Yadav <t.yadav@samsung.com>

raveendra-karu committed:
Updated namespace

Author: Tushant Yadav <t.yadav@samsung.com>

raveendra-karu committed:
Rebase to latest code.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

raveendra-karu committed:
Address LINT errors.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

yufengwangca committed:
Fix compile erros to unblock CI

yufengwangca committed:
Fix compile error
